### PR TITLE
8322477: order of subclasses in the permits clause can differ between compilations

### PR DESCRIPTION
**Pull request by GoeLin**
I backport this for parity with 21.0.4-oracle.

src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java
src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java
I had to resolve because indentation is different. Changed coding is the same.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8322477](https://bugs.openjdk.org/browse/JDK-8322477) needs maintainer approval

### Issue
 * [JDK-8322477](https://bugs.openjdk.org/browse/JDK-8322477): order of subclasses in the permits clause can differ between compilations (**Bug** - P4 - Approved)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk21u-dev.git pull/310/head:pull/310` \
`$ git checkout pull/310`

Update a local copy of the PR: \
`$ git checkout pull/310` \
`$ git pull https://git.openjdk.org/jdk21u-dev.git pull/310/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 310`

View PR using the GUI difftool: \
`$ git pr show -t 310`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk21u-dev/pull/310.diff">https://git.openjdk.org/jdk21u-dev/pull/310.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk21u-dev/pull/310#issuecomment-1973411937)

### CODE REVIEW

**Comment by shipilev** [comment]
I am surprised we have identing conflicts. Shouldn't we do the backport of https://bugs.openjdk.org/browse/JDK-8319987 first?

**Comment by GoeLin** [comment]
Hi, 
yes, JDK-8319987 might make Types.java clean.
in Attr.java, the indentation was changed by https://bugs.openjdk.org/browse/JDK-8316470: Incorrect error location for "invalid permits clause" depending on file order.
I don't think we should backport all dependencies just to avoid trivial conflicts. I chose this one to keep track with Oracle.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java [hunk 1]
```diff
@@ -5492,7 +5492,7 @@ void attribClass(ClassSymbol c) throws CompletionFailure {
 
                 if (!c.type.isCompound()) {
                     for (ClassSymbol supertypeSym : sealedSupers) {
-                        if (!supertypeSym.permitted.contains(c.type.tsym)) {
+                       if (!supertypeSym.isPermittedSubclass(c.type.tsym)) {
```
Comment (TheRealMDoerr) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java:5495]: Indentation is off by one.

**Review by TheRealMDoerr — COMMENTED** [review]
Looks basically good. Only one minor thing below. https://bugs.openjdk.org/browse/JDK-8319987 would only resolve one trivial diff. Is that planned to be backported for 21.0.4, too? If nobody wants to backport that one very soon, I don't think we have to wait.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java (hunk 1 above)
Comment (GoeLin) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java]: Thanks for the review. Fixed.

**Review by TheRealMDoerr — APPROVED** [review]

**Pull request closed without merging**